lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+			defer bc.snapshot.StartAncestralRecordsFlush(true)
+			glog.V(2).Infof("upsertBlockAndBlockNodeToDB: Preparing snapshot flush")
+		}
+		// TODO: Do we want to write the full block once.
+		// Store the new block in the db under the
+		//   <blockHash> -> <serialized block>
+		// index.
+		// TODO: In the archival mode, we'll be setting ancestral entries for the block reward. Note that it is
+		// 	set in PutBlockWithTxn. Block rewards are part of the state, and they should be identical to the ones
+		// 	we've fetched during Hypersync. Is there an edge-case where for some reason they're not identical? Or
+		// 	somehow ancestral records get corrupted?
+		if storeFullBlock {
+			if innerErr := PutBlockWithTxn(txn, bc.snapshot, block); innerErr != nil {
+				return errors.Wrapf(innerErr, "upsertBlockAndBlockNodeToDB: Problem calling PutBlock")
+			}
+		}
+
+		// TODO: if storeFullBlock = false, then we should probably remove the block from the DB? This can
+		// happen if we had a block stored in the DB but then determined that it would have failed validation.
+		// We would need to evict the block from the DB in that case.
+
+		// Store the new block's node in our node index in the db under the
+		//   <height uin32, blockHash BlockHash> -> <node info>
+		// index.
+		if innerErr := PutHeightHashToNodeInfoWithTxn(txn, bc.snapshot, blockNode, false /*bitcoinNodes*/); innerErr != nil {
+			return errors.Wrapf(innerErr, "upsertBlockAndBlockNodeToDB: Problem calling PutHeightHashToNodeInfo before validation")
+		}
+
+		// Notice we don't call PutBestHash or PutUtxoOperationsForBlockWithTxn because we're not
+		// affecting those right now.
+
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "upsertBlockAndBlockNodeToDB: Problem putting block in db: ")
+	}
+	return nil
+}
+
+// tryApplyNewTip attempts to apply the new tip to the best chain. It will do the following:
+// 1. Check if we should perform a reorg. If so, it will handle the reorg. If reorging causes an error, return false and error.
+// 2. Check if the incoming block extends the chain tip after reorg. If not, return false and nil
+// 3. If the incoming block extends the chain tip, we can apply it by calling addBlockToBestChain. Return true and nil.
+func (bc *Blockchain) tryApplyNewTip(blockNode *BlockNode, currentView uint64, lineageFromCommittedTip []*BlockNode) (_appliedNewTip bool, _err error) {
+
+	// Check if the incoming block extends the chain tip. If so, we don't need to reorg
+	// and can just add this block to the best chain.
+	chainTip := bc.GetBestChainTip()
+	if chainTip.Hash.IsEqual(blockNode.Header.PrevBlockHash) {
+		bc.addBlockToBestChain(blockNode)
+		return true, nil
+	}
+	// Check if we should perform a reorg here.
+	// If we shouldn't reorg AND the incoming block doesn't extend the chain tip, we know that
+	// the incoming block will not get applied as the new tip.
+	if !bc.shouldReorg(blockNode, currentView) {
```

Comment:
This shouldReorg thing was confusing me a bit. Consider the following examples:
* `bestChain = A(view=1) -> B(view=2, uncommitted) -> C(view=3, uncommitted)` and now we have block `C'(view=3, uncommitted)`. I think that we would always reorg to C' in this case, which I think is fine.
* But what if we have `bestChain = A(view=1)` and we have `B(view=2, uncommitted)` and `C(view=3, uncommitted)` such that B and C aren't in the bestChain yet. I think this function would still count that as a reorg even though we're just extending the bestChain by two.

I think the latter scenario is still fine, it's just a little weird. But I also think it probably can't happen in practice the way we call this. I'm still reviewing.

Reply:
I don't think we could have a situation where we didn't add B and C to the BestChain.

Reply:
Confirmed you can't connect C without B already being in bestChain. This is because of the way we process orphans at the end of processBlockPoS.